src/aspire/basis/polar_2d.py

```diff
@@ -0,0 +1,122 @@
+        Initialize an object for the 2D polar Fourier grid class
+
+        :param size: The shape of the vectors for which to define the grid.
+            Currently only square images are supported.
+        :param nrad: The number of points in the radial dimension.
+        :param ntheta: The number of points in the angular dimension.
+        """
+
+        ndim = len(size)
+        ensure(ndim == 2, 'Only two-dimensional grids are supported.')
+        ensure(len(set(size)) == 1, 'Only square domains are supported.')
+
+        self.nrad = nrad
+        if nrad is None:
+            self.nrad = self.nres // 2
+
+        self.ntheta = ntheta
+        if ntheta is None:
+            # try to use the same number as Fast FB basis
+            self.ntheta = 8 * self.nrad
+
+        super().__init__(size)
+
+    def _build(self):
+        """
+        Build the internal data structure to 2D polar Fourier grid
+        """
+        logger.info('Represent 2D image in a polar Fourier grid')
+
+        self.count = self.nrad * self.ntheta
+        self._sz_prod = self.sz[0] * self.sz[1]
+
+        # precompute the basis functions in 2D grids
+        self.freqs = self._precomp()
+
+    def _precomp(self):
+        """
+        Precomute the polar Fourier grid
+        """
+        omega0 = 2 * np.pi / (2 * self.nrad - 1)
+        dtheta = 2 * np.pi / self.ntheta
+
+        # Note: The size of ntheta can be reduced by half if only evaluate_t function is needed as
+        # in the Matlab implementation, np.zeros((2, self.nrad * self.ntheta//2)).
+        freqs = np.zeros((2, self.nrad * self.ntheta))
+        for i in range(self.ntheta):
+            freqs[0, i * self.nrad: (i + 1) * self.nrad] = np.arange(self.nrad) * np.sin(i * dtheta)
+            freqs[1, i * self.nrad: (i + 1) * self.nrad] = np.arange(self.nrad) * np.cos(i * dtheta)
+
+        freqs *= omega0
+        return freqs
+
+    def evaluate(self, v):
+        """
+        Evaluate coefficients in standard 2D coordinate basis from those in polar Fourier basis
+
+        :param v: A coefficient vector (or an array of coefficient vectors)
+            in polar Fourier basis to be evaluated. The first two dimensions must equal
+            `self.nrad` and `self.ntheta`.
+        :return x: The evaluation of the coefficient vector(s) `x` in standard 2D
+            coordinate basis. This is an array whose first two dimensions equal `self.sz`
+            and the remaining dimensions correspond to dimensions two and higher of `v`.
+        """
+        # TODO: need check the normalization factor and develop unit test
+        v, sz_roll = unroll_dim(v, 3)
+        v = m_reshape(v, (self.nrad*self.ntheta, v.shape[2]))
+        nimgs = v.shape[1]
+        # finufftpy require it to be aligned in fortran order
+        pf = np.empty((self._sz_prod, nimgs), dtype='complex128', order='F')
+        finufftpy.nufft2d1many(self.freqs[0], self.freqs[1], v, 1, 1e-15, self.sz[0], self.sz[1], pf)
```

Comment:
Should be able to only give half of the `v` coefficients here (just like in `evaluate_t`).

Reply:
I checked the FB2D basis classes. It looks that the other half is still needed for evaluating.

Reply:
Even if you cast the output to real? Also, `pf` is not a polar Fourier transform here, so it needs a different name.

Reply:
To satisfy the adjoint operator, I can not cast the output to real. So probably I will left it as it.  How about change `pf` to `x` as in other basis class?

Reply:
> To satisfy the adjoint operator, I can not cast the output to real.

But `evaluate_t` only calculates the Fourier transform for half the frequencies, so its input must be real. Its adjoint must therefore have real outputs.

> How about change pf to x as in other basis class?

Yes that sounds good.

Reply:
For `evaluate_t,` the input x vector is real, the the first half of output v vector is obtained from the FFT transform and the second half is obtained by the conjugate property. For the `evaluate,` the input v vector is complex, but the output x should be real. It looks that i can not just double the real value from the first half frequencies such as 
`finufftpy.nufft2d1many(self.freqs[0, :half_size], self.freqs[1, :half_size], v[:half_size, ...], 1, 1e-15, self.sz[0], self.sz[1], x)`.  In FB basis class, it looks that both - and + parts have to be evaluated.

Reply:
Unrelated, but can I ask why we don't use the nufft class structure here? We do use it elsewhere.  This would allow the nufft calls can be automatically swapped out depending on the backend.  As this is written it will ignore user configuration selecting a different library etc, and also will not work in the case finufft is N/A.  Is this just for testing? Thanks

Reply:
The implementation of passing of many frames is not in the develop branch yet. I did this way so that I should not repeat @garrettwrong's work for including the new functionality of finufft. We can easily to switch to when the new interface of many frames is ready to use.

Reply:
_But this is an active request to put this code into the `development` branch?_

The nufft class was something Vineet wrote a long time ago.  While I appreciate saying you do not want to duplicate the `many` method work, you should know using many is only an optimization, it is not required to call finufft or the other nufft libraries that way.  Also, we both know your code was written long before my recent work _for the hackathon_ ;)

Many does perform better, but clearly since we ported code inside out, we're not really concerned with performance at this time...  Using the single transform setup, execution, and destructor does exist packaged as the master's nufft class, and is what we use elsewhere, like under the cov2d example.  It is actually the entire setup, execute, and destroy that I am concerned about, not the optimization.  To implement a many _optimization_ when using the class structure is a few line change later.  To survey, catalog, and re-implement code we just "ported" to use the ASPIRE python class structures, seems like a core concern to me.  I guess I don't understand the process we are using for new code.  Are we intending to rewrite it several times in the development branch?

As I see it, the problem with implementing everything adhoc again in the python code, is that someone else will need to come along and rewrite the ported codes (perhaps several times it seems) before they are production worthy... so it is repeating work, just in the future. That is why I asked if this was just for testing purposes.

Imagine some user/developer doesn't have finufft, or wants to use a different library, turns the ASPIRE configuration to a different library and none of this code works... bad times.

Reply:
Yes. Many parts of code in the development branch can be reimplemented in a better way.  We are trying to reach a minimum set of modules using our new framework which can perform the whole pipeline of reconstruction. We would like to reach this in the summer but decide to postpone it to the end of 2020 for a reasonable estimation.  Currently the CPU version of finufft was installed through PyPI package. Of course we might change it later.  We can discuss these in the meeting.

Reply:
> It looks that i can not just double the real value from the first half frequencies

I don't quite understand why this cannot be done? Taking the real part after performing the IFT corresponds to symmetrizing before the frequencies, then taking the IFT. I will take a stab.

Regarding the direct call to the Finufft module, part of this is my fault. To get some reasonable speeds for the covariance estimation, I started doing this instead of (properly) implementing the many interface in the Nufft wrapper classes. Why don't we switch to using the wrapper classes now (at a certain loss in performance) and prioritize getting those wrappers to use the many interface as soon as possible?